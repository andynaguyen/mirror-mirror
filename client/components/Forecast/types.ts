import { QueryProps } from 'react-apollo';
import { ForecastData } from '../../../common/types';

interface ForecastProps extends ApolloWrapperProps {
  data?: QueryProps<ForecastData>;
  forecast: ForecastData;
}

interface ApolloWrapperProps {
  visible: boolean;
}

type ForecastResponse = {
  forecast: ForecastData;
};

export { ForecastProps, ApolloWrapperProps, ForecastResponse };
