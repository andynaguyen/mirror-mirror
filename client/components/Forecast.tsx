import axios from 'axios';
import * as React from 'react';
import styled from 'styled-components';
import withRefresh from '../refresher';

const REFRESH_INTERVAL = 1000 * 60 * 60; // 1 hour

// Maps the icon name from the API to the proper weather icon class name.
const ICON_MAP = {
  'clear-day': 'wi wi-day-sunny',
  'clear-night': 'wi wi-night-clear',
  'cloudy': 'wi wi-cloudy',
  'default': 'wi wi-na',
  'fog': 'wi wi-fog',
  'hail': 'wi wi-hail',
  'partly-cloudy-day': 'wi wi-day-cloudy',
  'partly-cloudy-night': 'wi wi-night-cloudy',
  'rain': 'wi wi-rain',
  'sleet': 'wi wi-sleet',
  'snow': 'wi wi-snow',
  'thunderstorm': 'wi wi-thunderstorm',
  'tornado': 'wi wi-tornado',
  'wind': 'wi wi-strong-wind',
};

interface ForecastProps {
  icon: string;
  precipitation: number;
  summary: string;
  temperature: number;
  wind: number;
  visible: boolean;
}

const Temperature = styled.div`
  font-family: 'Raleway';
  font-size: 48px;
  float: left;
`;

const Summary = styled.span`
  font-size: 20px;
  margin-top: 4px;
  margin-bottom: 4px;
`;

const PlainText = styled.span`
  font-family: 'Raleway';
`;

const WeatherIcon = styled.i`
  font-size: 64px;
  margin-left: 30px;
`;

const MiniWeatherIcon = styled.i`
  font-size: 24px;
  margin-right: 12px;
`;

const Forecast: React.SFC<ForecastProps> = ({ icon, precipitation, summary, temperature, wind, visible }) => {
  const iconClass = ICON_MAP[icon] || ICON_MAP.default;
  return visible ? (
    <>
      <div>
        <Temperature>
          {temperature} <i className="wi wi-fahrenheit" />
        </Temperature>
        <WeatherIcon className={iconClass} />
      </div>
      <div>
        <Summary>{summary}</Summary>
      </div>
      <div>
        <MiniWeatherIcon className="wi wi-umbrella" />
        <PlainText>{precipitation}</PlainText>%
      </div>
      <div>
        <MiniWeatherIcon className="wi wi-strong-wind" />
        <PlainText>{wind}</PlainText> mph
      </div>
    </>
  ) : null;
};

const defaultState = {
  icon: 'default',
  precipitation: 0,
  summary: 'N/A',
  temperature: 0,
  wind: 0,
};

const options = {
  defaultState,
  interval: REFRESH_INTERVAL,
  refresh: async () => {
    try {
      const forecast = await axios.get('/api/forecast');
      return forecast.data;
    } catch (error) {
      console.error(error);
      return defaultState;
    }
  },
};

/**
 * A component showing the weather forecast that updates every hour.
 */
const EnhancedForecast = withRefresh(options, Forecast);
export default EnhancedForecast;
