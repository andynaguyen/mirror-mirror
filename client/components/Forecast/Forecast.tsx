import * as React from 'react';
import styled from 'styled-components';
import { ForecastProps } from './types';

// Maps the icon name from the API to the proper weather icon class name.
const ICON_MAP = {
  'clear-day': 'wi wi-day-sunny',
  'clear-night': 'wi wi-night-clear',
  cloudy: 'wi wi-cloudy',
  default: 'wi wi-na',
  fog: 'wi wi-fog',
  hail: 'wi wi-hail',
  'partly-cloudy-day': 'wi wi-day-cloudy',
  'partly-cloudy-night': 'wi wi-night-cloudy',
  rain: 'wi wi-rain',
  sleet: 'wi wi-sleet',
  snow: 'wi wi-snow',
  thunderstorm: 'wi wi-thunderstorm',
  tornado: 'wi wi-tornado',
  wind: 'wi wi-strong-wind',
};

const Temperature = styled.div`
  font-family: 'Raleway';
  font-size: 48px;
  float: left;
`;

const Summary = styled.span`
  font-size: 20px;
  margin-top: 4px;
  margin-bottom: 4px;
`;

const PlainText = styled.span`
  font-family: 'Raleway';
`;

const WeatherIcon = styled.i`
  font-size: 64px;
  margin-left: 30px;
`;

const MiniWeatherIcon = styled.i`
  font-size: 24px;
  margin-right: 12px;
`;

export default ({ forecast, visible }: ForecastProps) => {
  const { icon, precipitation, summary, temperature, wind } = forecast;
  const iconClass = ICON_MAP[icon] || ICON_MAP.default;
  return visible ? (
    <>
      <div>
        <Temperature>
          {temperature} <i className="wi wi-fahrenheit" />
        </Temperature>
        <WeatherIcon className={iconClass} />
      </div>
      <div>
        <Summary>{summary}</Summary>
      </div>
      <div>
        <MiniWeatherIcon className="wi wi-umbrella" />
        <PlainText>{precipitation}</PlainText>%
      </div>
      <div>
        <MiniWeatherIcon className="wi wi-strong-wind" />
        <PlainText>{wind}</PlainText> mph
      </div>
    </>
  ) : null;
};
