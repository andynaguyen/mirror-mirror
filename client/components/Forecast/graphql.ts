import { graphql } from 'react-apollo';
import gql from 'graphql-tag';
import { ForecastResponse, ApolloWrapperProps } from './types';

const query = gql`
  query {
    forecast {
      icon
      precipitation
      summary
      temperature
      wind
    }
  }
`;

export default graphql<ForecastResponse, ApolloWrapperProps>(query, {
  options: {
    pollInterval: 1000 * 60 * 60, // 1 hour
    fetchPolicy: 'network-only',
  },
  props: ({ data: { forecast, error } }) => ({
    forecast: {
      icon: error || !forecast ? 'default' : forecast.icon,
      precipitation: error || !forecast ? -1 : forecast.precipitation,
      summary: error || !forecast ? 'N/A' : forecast.summary,
      temperature: error || !forecast ? -1 : forecast.temperature,
      wind: error || !forecast ? -1 : forecast.wind,
    },
  }),
});
