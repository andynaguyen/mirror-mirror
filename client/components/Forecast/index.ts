import Forecast from './Forecast';
import withGraphql from './graphql';

/**
 * A component showing the weather forecast that updates every hour.
 */
const EnhancedForecast = withGraphql(Forecast);
export default EnhancedForecast;
