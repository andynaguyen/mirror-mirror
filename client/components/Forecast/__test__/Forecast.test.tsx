import 'jest-styled-components';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import Forecast from '../Forecast';

describe('<Forecast />', () => {
  const forecastData = {
    icon: 'default',
    precipitation: 9.3,
    summary: 'Showers',
    temperature: 56.7,
    wind: 2,
  };
  it('should render the contents of the forecast component', () => {
    const component = renderer.create(<Forecast forecast={forecastData} visible />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render nothing', () => {
    const component = renderer.create(<Forecast forecast={forecastData} visible={false} />);
    const tree = component.toJSON();
    expect(tree).toBeNull();
  });
});
