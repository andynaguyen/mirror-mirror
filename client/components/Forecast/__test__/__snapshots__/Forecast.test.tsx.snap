// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Forecast /> should render the contents of the forecast component 1`] = `
Array [
  .c0 {
  font-family: 'Raleway';
  font-size: 48px;
  float: left;
}

.c1 {
  font-size: 64px;
  margin-left: 30px;
}

<div>
    <div
      className="c0"
    >
      56.7
       
      <i
        className="wi wi-fahrenheit"
      />
    </div>
    <i
      className="wi wi-na c1"
    />
  </div>,
  .c0 {
  font-size: 20px;
  margin-top: 4px;
  margin-bottom: 4px;
}

<div>
    <span
      className="c0"
    >
      Showers
    </span>
  </div>,
  .c1 {
  font-family: 'Raleway';
}

.c0 {
  font-size: 24px;
  margin-right: 12px;
}

<div>
    <i
      className="wi wi-umbrella c0"
    />
    <span
      className="c1"
    >
      9.3
    </span>
    %
  </div>,
  .c1 {
  font-family: 'Raleway';
}

.c0 {
  font-size: 24px;
  margin-right: 12px;
}

<div>
    <i
      className="wi wi-strong-wind c0"
    />
    <span
      className="c1"
    >
      2
    </span>
     mph
  </div>,
]
`;
